internal/moduletest/mocking/fill.go

```diff
@@ -0,0 +1,229 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package mocking
+
+import (
+	"fmt"
+
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/convert"
+
+	"github.com/hashicorp/terraform/internal/configs/configschema"
+)
+
+// FillAttribute makes the input value match the specified attribute by adding
+// attributes and/or performing conversions to make the input value correct.
+//
+// It is similar to FillType, except it accepts attributes instead of types.
+func FillAttribute(in cty.Value, attribute *configschema.Attribute) (cty.Value, error) {
+	return fillAttribute(in, attribute, cty.Path{})
+}
+
+func fillAttribute(in cty.Value, attribute *configschema.Attribute, path cty.Path) (cty.Value, error) {
+	if attribute.NestedType != nil {
+
+		// Then the in value must be an object.
+		if !in.Type().IsObjectType() {
+			return cty.NilVal, path.NewErrorf("incompatible types; expected object type, found %s", in.Type().FriendlyName())
+		}
+
+		switch attribute.NestedType.Nesting {
+		case configschema.NestingSingle, configschema.NestingGroup:
+			children := make(map[string]cty.Value)
+			for name, attribute := range attribute.NestedType.Attributes {
+				if in.Type().HasAttribute(name) {
+					child, err := fillAttribute(in.GetAttr(name), attribute, path.GetAttr(name))
+					if err != nil {
+						return cty.NilVal, err
+					}
+					children[name] = child
+					continue
+				}
+
+				children[name] = GenerateValueForAttribute(attribute)
+			}
+			if len(children) == 0 {
+				return cty.EmptyObjectVal, nil
+			}
+			return cty.ObjectVal(children), nil
+		case configschema.NestingSet:
+			return cty.SetValEmpty(attribute.ImpliedType().ElementType()), nil
+		case configschema.NestingList:
+			return cty.ListValEmpty(attribute.ImpliedType().ElementType()), nil
+		case configschema.NestingMap:
+			return cty.MapValEmpty(attribute.ImpliedType().ElementType()), nil
+		default:
+			panic(fmt.Errorf("unknown nesting mode: %d", attribute.NestedType.Nesting))
+		}
+	}
+
+	return FillType(in, attribute.Type)
+}
+
+// FillType makes the input value match the target type by adding attributes
+// directly to it or to any nested objects. Essentially, this is a "safe"
+// conversion between two objects.
+//
+// This function can error if one of the embedded types within value doesn't
+// match the type expected by target.
+//
+// If the supplied value isn't an object (or a map that can be treated as an
+// object) then a normal conversion is attempted from value into target.
+//
+// Superfluous attributes within the supplied value (ie. attributes not
+// mentioned by the target type) are dropped without error.
+func FillType(in cty.Value, target cty.Type) (cty.Value, error) {
+	return fillType(in, target, cty.Path{})
+}
+
+func fillType(in cty.Value, target cty.Type, path cty.Path) (cty.Value, error) {
+	// If we're targeting an object directly, and we have an object then we just
+	// need
```

Comment:
I think this comment trails off.

Given that the logic is pretty much identical for a map input type, you might want one comment to cover both (or a comment above the map test to refer back to this one).

Reply:
Done! Thanks!